File: 2016/luke15.hs
import Data.List

main :: IO ()
main = print . length $ lookAndSay 50 "1111321112321111"

lookAndSay :: Int -> String -> String
lookAndSay 0 nums = nums
lookAndSay i nums = lookAndSay (i-1) expanded
   where expand n = show (length n) ++ [head n]
         expanded = concatMap expand $ group nums

File: 2016/luke15.scala
object Luke15 {
	def lookAndSay(i: Int, nums: String): String =
		if (i == 0) nums
		else lookAndSay(i-1, expand(nums).mkString)

	def expand(nums: String) =
		nums.foldRight(List.empty[(Int, String)])((n, acc) => acc match {
			case (h :: t) if h._2 == n.toString => (h._1 + 1, h._2) :: t
			case _ => (1, n.toString) :: acc
		}) flatMap (_.productIterator.toList)

	def main(args:Array[String]) = println(lookAndSay(50, "1111321112321111").length)
}
